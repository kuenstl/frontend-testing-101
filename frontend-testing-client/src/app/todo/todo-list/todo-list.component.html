<section class="todoapp">
  <header class="header">
    <h1>Todos</h1>

    <input class="new-todo" placeholder="What needs to be done?" autofocus="" 
      [(ngModel)]="newTodoText" (keyup.enter)="addTodo()">

  </header>

  <section class="main">
    <ul class="todo-list">

      <li *ngFor="let todo of todos" 
        [class.completed]="todo.completed">

        <div class="view">
          <input class="toggle" type="checkbox" 
            (click)="toggleCompletion(todo._id)" [checked]="todo.completed">
          <label>{{ todo.text }}</label>
          <button class="destroy" (click)="delete(todo)"></button>
        </div>

      </li>
    </ul>
  </section>

</section>