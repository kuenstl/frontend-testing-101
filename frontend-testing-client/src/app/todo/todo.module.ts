import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';

import { TodoComponent } from './todo.component';
import { TodoService } from './todo.service';

@NgModule({
  declarations: [
    TodoComponent
  ],
  imports: [
    BrowserModule,
    CommonModule,
    FormsModule,
  ],
  providers: [TodoService],
  exports: [
    TodoComponent
  ]
})
export class TodoModule { }
