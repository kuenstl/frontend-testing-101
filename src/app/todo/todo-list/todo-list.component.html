<section class="todoapp">
  <header class="header">
    <h1>Todos</h1>

    <input class="new-todo" placeholder="What needs to be done?" autofocus=""
      [(ngModel)]="newTodoText" (keyup.enter)="addTodo()">

  </header>

  <section class="main">
    <ul class="todo-list">

      <ft-todo-item *ngFor="let todo of todos"
        [todo]="todo"
        (onCompleted)="toggleCompletion(todo)"
        (onDeleted)="delete(todo)"></ft-todo-item>

    </ul>
  </section>

</section>
