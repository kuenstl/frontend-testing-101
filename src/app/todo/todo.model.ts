export class Todo {
  completed: boolean;
  editing: boolean;

  constructor(public title: string) {
    this.completed = false;
    this.editing = false;
    this.title = title.trim();
  }
}
