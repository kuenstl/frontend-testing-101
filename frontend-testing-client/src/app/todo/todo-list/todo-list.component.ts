import { Component, OnInit } from '@angular/core';

import { Todo } from '../todo.model';
import { TodoService } from '../todo.service';

@Component({
  selector: 'ft-todo-list',
  templateUrl: './todo-list.component.html'
})
export class TodoListComponent implements OnInit {
  todos: Todo[];
  newTodoText = '';

  constructor(private todoService: TodoService) {}

  ngOnInit() {
    this.getTodos();
  }

  getTodos(): void {
    this.todoService.getTodos().subscribe(todos => (this.todos = todos));
  }

  toggleCompletion(todo: Todo) {
    todo.completed = !todo.completed;
    this.todoService.updateTodo(todo).subscribe(todo => {
      const idx = this.todos.findIndex(t => todo._id === t._id);
      this.todos[idx] = todo;
    });
  }

  delete(todo: Todo): void {
    this.todos = this.todos.filter(t => t !== todo);
    this.todoService.deleteTodo(todo).subscribe();
  }

  addTodo() {
    this.newTodoText = this.newTodoText.trim();

    if (!this.newTodoText) {
      return;
    }
    this.todoService
      .addTodo(new Todo(this.newTodoText))
      .subscribe(todo => this.todos.push(todo));

    this.newTodoText = '';
  }
}
