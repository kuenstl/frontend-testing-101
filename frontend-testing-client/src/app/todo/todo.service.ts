import {Todo} from './todo.model';

const localStorageKey = 'ngx-todos';

export class TodoService {
  todos: Array<Todo>;

  constructor() {
    let persistedTodos =
        JSON.parse(localStorage.getItem(localStorageKey) || '[]');

    this.todos =
        persistedTodos.map((todo: {title: string, completed: boolean}) => {
          let newTodo = new Todo(todo.title);
          newTodo.completed = todo.completed;
          return newTodo;
        });
  }

  allCompleted() {
    return this.todos.length === this.getCompleted().length;
  }

  setAllTo(completed: boolean) {
    this.todos.forEach((todo: Todo) => todo.completed = completed);
    this.update();
  }

  removeCompleted() {
    this.todos = this.getWithCompleted(false);
    this.update();
  }

  getRemaining() {
    return this.getWithCompleted(false);
  }

  getCompleted() {
    return this.getWithCompleted(true);
  }

  toggleCompletion(todo: Todo) {
    todo.completed = !todo.completed;
    this.update();
  }

  remove(todo: Todo) {
    this.todos.splice(this.todos.indexOf(todo), 1);
    this.update();
  }

  add(title: string) {
    this.todos.push(new Todo(title));
    this.update();
  }

  private update() {
    localStorage.setItem(localStorageKey, JSON.stringify(this.todos));
  }

  private getWithCompleted(completed: boolean) {
    return this.todos
      .filter((todo: Todo) => todo.completed === completed);
  }
}