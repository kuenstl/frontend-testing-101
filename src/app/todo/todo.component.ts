import { Component } from '@angular/core';

import { Todo } from './todo.model';
import { TodoService } from './todo.service';

@Component({
  selector: 'ft-todo', 
  templateUrl: './todo.component.html'
})
export class TodoComponent {
  
  newTodoText = '';

  constructor(private todoService: TodoService) { }

  stopEditing(todo: Todo, editedTitle: string) {
    todo.title = editedTitle;
    todo.editing = false;
  }

  cancelEditingTodo(todo: Todo) {
    todo.editing = false;
  }

  updateEditingTodo(todo: Todo, editedTitle: string) {
    editedTitle = editedTitle.trim();
    todo.editing = false;

    if (editedTitle.length === 0) {
      return this.todoService.remove(todo);
    }

    todo.title = editedTitle;
  }

  editTodo(todo: Todo) {
    todo.editing = true;
  }

  removeCompleted() {
    this.todoService.removeCompleted();
  }

  toggleCompletion(todo: Todo) {
    this.todoService.toggleCompletion(todo);
  }

  remove(todo: Todo) {
    this.todoService.remove(todo);
  }

  addTodo() {
    if (this.newTodoText.trim().length) {
      this.todoService.add(this.newTodoText);
      this.newTodoText = '';
    }
  }
}