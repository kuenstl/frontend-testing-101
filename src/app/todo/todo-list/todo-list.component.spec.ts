import { async, TestBed, fakeAsync, tick } from '@angular/core/testing';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';

import { TodoItemComponent } from '../todo-item/todo-item.component';
import { TodoListComponent } from './todo-list.component';
import { TodoService } from '../todo.service';
import { Todo } from '../todo.model';
import { By } from '@angular/platform-browser';
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/observable/from';

const firstTestTodo = new Todo('Write unit tests');
const secondTestTodo = new Todo('Write more unit tests');

describe('TodoListComponent', () => {
  let fixture;
  let component;
  let todoService;
  let addTodoSpy;

  beforeEach(() => {
    const todoServiceStub = {
      getTodos() {
        return Observable.from([[firstTestTodo, secondTestTodo]]);
      },

      addTodo(todo) {
        return Observable.from([todo]);
      }
    };

    TestBed.configureTestingModule({
      imports: [CommonModule, FormsModule],
      declarations: [TodoItemComponent, TodoListComponent],
      providers: [{ provide: TodoService, useValue: todoServiceStub }]
    }).compileComponents();

    fixture = TestBed.createComponent(TodoListComponent);
    component = fixture.debugElement.componentInstance;
    todoService = fixture.debugElement.injector.get(TodoService);

    addTodoSpy = jest.spyOn(todoService, 'addTodo');

    fixture.detectChanges();
  });

  it('should create the component', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should show existing todos', () => {
    expect(component.todos).toContain(firstTestTodo);
    expect(component.todos).toContain(secondTestTodo);
    expect(fixture).toMatchSnapshot();
  });

  it('should add a todo', fakeAsync(() => {
    const newTestTodo = new Todo('Write another unit test');
    component.newTodoText = newTestTodo.text;
    fixture.detectChanges();

    component.addTodo(newTestTodo);

    fixture.detectChanges();
    tick();
    fixture.detectChanges();

    expect(addTodoSpy).toHaveBeenCalledWith(newTestTodo);
    expect(fixture).toMatchSnapshot();
  }));
});
