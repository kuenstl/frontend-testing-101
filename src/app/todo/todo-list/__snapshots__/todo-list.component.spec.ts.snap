// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListComponent should add a todo 1`] = `
<ft-todo-list
  newTodoText=""
  todoService={[Function Object]}
  todos={[Function Array]}
>
  <section
    class="todoapp"
  >
    
    
    <header
      class="header"
    >
      
      
      <h1>
        Todos
      </h1>
      
  
      
      <input
        autofocus=""
        class="new-todo ng-untouched ng-pristine ng-valid"
        ng-reflect-model=""
        placeholder="What needs to be done?"
      />
      
  
    
    </header>
    
  
    
    <section
      class="main"
    >
      
      
      <ul
        class="todo-list"
      >
        
  
        
        
        <ft-todo-item
          ng-reflect-todo="[object Object]"
        >
          <li>
            
    
            <div
              class="view"
            >
              
      
              <input
                class="toggle"
                type="checkbox"
              />
              
      
              <label>
                Write unit tests
              </label>
              
      
              <button
                class="destroy"
              />
              
    
            </div>
            
  
          </li>
          
  
        </ft-todo-item>
        <ft-todo-item
          ng-reflect-todo="[object Object]"
        >
          <li>
            
    
            <div
              class="view"
            >
              
      
              <input
                class="toggle"
                type="checkbox"
              />
              
      
              <label>
                Write more unit tests
              </label>
              
      
              <button
                class="destroy"
              />
              
    
            </div>
            
  
          </li>
          
  
        </ft-todo-item>
        <ft-todo-item
          ng-reflect-todo="[object Object]"
        >
          <li>
            
    
            <div
              class="view"
            >
              
      
              <input
                class="toggle"
                type="checkbox"
              />
              
      
              <label>
                Write another unit test
              </label>
              
      
              <button
                class="destroy"
              />
              
    
            </div>
            
  
          </li>
          
  
        </ft-todo-item>
        
  
      
      </ul>
      
    
    </section>
    
  
  
  </section>
  
</ft-todo-list>
`;

exports[`TodoListComponent should create the component 1`] = `
<ft-todo-list
  newTodoText=""
  todoService={[Function Object]}
  todos={[Function Array]}
>
  <section
    class="todoapp"
  >
    
    
    <header
      class="header"
    >
      
      
      <h1>
        Todos
      </h1>
      
  
      
      <input
        autofocus=""
        class="new-todo ng-untouched ng-pristine ng-valid"
        ng-reflect-model=""
        placeholder="What needs to be done?"
      />
      
  
    
    </header>
    
  
    
    <section
      class="main"
    >
      
      
      <ul
        class="todo-list"
      >
        
  
        
        
        <ft-todo-item
          ng-reflect-todo="[object Object]"
        >
          <li>
            
    
            <div
              class="view"
            >
              
      
              <input
                class="toggle"
                type="checkbox"
              />
              
      
              <label>
                Write unit tests
              </label>
              
      
              <button
                class="destroy"
              />
              
    
            </div>
            
  
          </li>
          
  
        </ft-todo-item>
        <ft-todo-item
          ng-reflect-todo="[object Object]"
        >
          <li>
            
    
            <div
              class="view"
            >
              
      
              <input
                class="toggle"
                type="checkbox"
              />
              
      
              <label>
                Write more unit tests
              </label>
              
      
              <button
                class="destroy"
              />
              
    
            </div>
            
  
          </li>
          
  
        </ft-todo-item>
        
  
      
      </ul>
      
    
    </section>
    
  
  
  </section>
  
</ft-todo-list>
`;

exports[`TodoListComponent should show existing todos 1`] = `
<ft-todo-list
  newTodoText=""
  todoService={[Function Object]}
  todos={[Function Array]}
>
  <section
    class="todoapp"
  >
    
    
    <header
      class="header"
    >
      
      
      <h1>
        Todos
      </h1>
      
  
      
      <input
        autofocus=""
        class="new-todo ng-untouched ng-pristine ng-valid"
        ng-reflect-model=""
        placeholder="What needs to be done?"
      />
      
  
    
    </header>
    
  
    
    <section
      class="main"
    >
      
      
      <ul
        class="todo-list"
      >
        
  
        
        
        <ft-todo-item
          ng-reflect-todo="[object Object]"
        >
          <li>
            
    
            <div
              class="view"
            >
              
      
              <input
                class="toggle"
                type="checkbox"
              />
              
      
              <label>
                Write unit tests
              </label>
              
      
              <button
                class="destroy"
              />
              
    
            </div>
            
  
          </li>
          
  
        </ft-todo-item>
        <ft-todo-item
          ng-reflect-todo="[object Object]"
        >
          <li>
            
    
            <div
              class="view"
            >
              
      
              <input
                class="toggle"
                type="checkbox"
              />
              
      
              <label>
                Write more unit tests
              </label>
              
      
              <button
                class="destroy"
              />
              
    
            </div>
            
  
          </li>
          
  
        </ft-todo-item>
        
  
      
      </ul>
      
    
    </section>
    
  
  
  </section>
  
</ft-todo-list>
`;
